Added cloudflare.com as default ip lookup

9cco opened the pull request:
Used shell built-in command -v to determine if dig is installed and if
it is, use it as default for getting the ip-address. Also made the use
of brackets consistent and updated the the way the data variable is
created before it is sent with curl to make it more readable in the
code.

timetoexpire commented:
Why? What is advantage of using `dig` over `curl` to do this. And even if use `dig` you still need to use `curl` when contacting cloudflare API. And as might not know ipv4.icanhazip.com (aka major.io) is owned by cloudflare.

Now taking your suggestion adding something like this would be good idea as `curl` is not installed by default on all distros.
```
  if ! command -v "curl" &> /dev/null; then
    echo "It is required to have curl installed (apt install curl)"
    exit 1
  fi
```

It same thing been doing with my folked version https://github.com/timetoexpire/cloudflare-ddns-updater `check_install_command curl`

9cco commented:
I think curl is using a http request to the server in order to get the ip. Dig on the other hand uses DNS requests and should be more secure in this sense. In the case ipv4.icanhazip.com becomes compromised, the curl requests is less secure I would think since the http protocol is more versatible, while the DNS protocol should be better suited for getting an ip address.

timetoexpire commented:
No it is not more secure using `dig`. If secury is the issue then `curl https` is lot more secure solution. It using HTTPS and not HTTP.
If I could get between your computer that using `dig` and (isp-internet) opendns.com I could very easy change/spoof ip that being reported back. As it just plane text.
If I could get between your computer that using `curl https` and (isp-internet) ipv4.icanhazip.com. I would need a valied cert for ipv4.icanhazip.com to be able change/spoof ip being reported back.

Have look at script for `curl` and it using HTTPS and not HTTP.  It will not drop back to HTTP. It check that that HTTPS is vailed for that domain and the cert has been signed by CA. If wanted it to not check for CA then it would have using ie `curl -k HTTPS`

As with ipv4.icanhazip.com being compromised, yep that could happen. That also could happen with opendns.com. But ipv4.icanhazip.com is owned and controlled by cloudflare. So why do trust cloudflare as your CND? You get my point?

`dig`
|   issue | status
|------|-----
handshake | no, udp
encripton | no, plane text
cert validation | no, plane text

`curl https`
|   issue | status
|------|-----
handshake | yes, tcp
encripton | yes, encripted
cert validation | yes, encription signed as vailed

9cco commented:
Thank you for clarifying. I didn't consider the MiTM angle of the security. What I was trying to address was that there was no control in the script of the input that curl provided. In PHP one has to be very careful with input like that as it can be used for code injection. My idea was that dig would provide more specific input, while curl could in general provide whatever input the remote site wanted to give.

I see you points though about the possibility of spoofing the ip, privacy of the transferred ip and additionally the issue of trust with that if we are already trusting cloudflare a bit why not continue using their service so we don't have to also trust opendns.

I tried to incorporate this into the script now by setting curl as the default ip provider program but also not allow any kind of input from curl but sanitize it through a regex which at least puts my heart a bit more at ease. I still added dig as a backup solution though in case curl is not installed on the system, which, as you mentioned earlier, could be a possibility.

timetoexpire commented:
It been possible to improve upon your **regex**.  I do think you idea of using **regex** is an good idea.

```
ipv4_regex () {
  ip="$1"
  echo -n "IP $ip "
  if [[ ! $ip =~ ^([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])$ ]]; then
    echo -n "[1] "
  fi
  if [[ ! $ip =~ [0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3} ]]; then
    echo -n "[2] "
  fi
  echo " "
}


datacsv="data.csv"
zeros=""
echo "y, x, zero, zeroes, boolean, T/F" >"$datacsv"
for (( zero=0; zero<3; zero++))
do
  echo "ZEROS $zero [$zeros]"
  for (( x=0; x<1000 ;x++))
  do
    y="$zeros""$x"
    out="$y, $x, $zero, $zeros, "

    if [[ $y =~ ^([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])$ ]]; then
     out+="1, T"
     echo -n "$y, "
    else
     out+="0, F"
    fi
    echo "$out" >>"$datacsv"
  done
  zeros+="0"
  echo " "
done

echo "My test triggered [1] your test has triggered [2]"
ipv4_regex 0.0.0.0
ipv4_regex 1.2.3.4
ipv4_regex 012.004.000.199
ipv4_regex 255.255.255.255
ipv4_regex 255.255.255.256
ipv4_regex 0.0.300.0
ipv4_regex 0..2.3
ipv4_regex 0.1
ipv4_regex 0.1f.2.3
```

Now the issue with `curl` has to be installed. `curl` also being used to contact cloudflare api. It not just for IPv4 address.
Another issue user issue with `dig` is the fact it not as simple to install like `curl` .  To install `curl` you just do ie `apt install curl`. To install `dig` it ie `apt install dnsutils` it not as obvious to end user.
Using `dig` still has the issue of MiTM.

All the best.

K0p1-Git commented:
Having previously worked with PHP, I get where you are coming from a security stand point. While it may be highly unlikely where the APIs get compromised, it should be factored in as well. Made some comments on your pull.

K0p1-Git commented:
Looks good to me. Have the script been tested @9cco ?

timetoexpire commented:
This what going on here. 
When check on system public IPv4 you checking if installed 'curl'. If 'curl' is installed use 'curl' .
If 'curl' isn't installed then check if 'dig' installed. If 'dig' is installed use 'dig'. 

What on if 'curl' or 'dig' not installed?
So you know 'curl' isn't installed. How do make contact with cloudflare api and slack api again as 'curl' isn't installed? Or have I missed something?

So what are 'dig' tangible advanages but to give Man in The Middle attacks an vector? That didn't exist when using 'curl https'. We still haven't got past issue which stated early on. Which why in my version my folked version I check if `curl` is installed and if not give some info to how to install it.

In my humble opinion the only advantageous thing here is sanitize of IPv4 by using **regex** .

9cco commented:
I think I tested it, but I'll try to make a double check later today.

@timetoexpire . Have you looked at the latest commit? The logic in the script right now is that no, we first check if `curl` is installed. But if it isn't installed then we quit with an error message with information about how to install it. I haven't gotten around to looking at your fork yet, but I'll try to find the time. From what I understand, that is what you were doing? Then we check the `regex` to see if the ip gained was valid. It is now that we have the possibility of using `dig`. If the regex isn't valid because e.g. the sites are down then we print a warning that we will try to use a more insecure *DNS* lookup through `dig`, test if dig is installed, if it is, try to use `dig`, test if the result is a valid ip with the same `regex` as before. Now if it isn't then we finally give up and exit saying we couldn't find the public ip. If the lookup gained from `dig` is valid, then we continue on our merry way.

With that being said. I think I get it now finally. If we have a Man in the Middle, then they could potentially also block access to the https sites forcing the client to use `dig` with the above mentioned logic then forcing them to use a malicious ip. So yes, I agree. `dig` shouldn't be an option because of this. Maybe I should change the name of the branch in order to reflect this? I'll look into how to do it in git. Maybe change it to 'regex sanitation'.

Btw. I also found that cloudflare has their own ip-lookup service hosted on 'https://cloudflare.com/cdn-cgi/trace' through *https*. I think this is the better solution than either to old ones and also `dig`. Since not everyone knows that 'api.ipify.org' is owned by cloudflare, having the lookup domain look like cloudflare would make it clear to the users of the script that we are trying to not use any 3rd party website, but by default only trust cloudflare.

timetoexpire commented:
I didn't know about https://www.cloudflare.com/cdn-cgi/trace
There is probably a better way of extracting IPv4

```
ip=$(curl -s -4 https://www.cloudflare.com/cdn-cgi/trace | grep "ip=")
ip=${ip#*"ip="}
echo "This is IPv4 $ip"
```

Why `curl -4`? So you get back only IPv4 if using IPv6. https://community.cloudflare.com/t/regarding-cdn-cgi-trace/239956

I would profer not to use `grep` `awk` `sed` but I can not at moment think of a way of doing it using only **regex**

9cco commented:
Ok so, I removed `dig` and added `curl -4 -s https://www.cloudflare.com/cdn-cgi/trace` as the default ip-lookup provider. I hope I at this point have addressed all your comments. In any case, thank you both for your feedback. 

@K0p1-Git I did unit tests for all commands and finally did an integration test to check that the whole thing worked as well, and it did. I think you hinted that we didn't really need the check for whether `curl` is installed so I removed that.

@timetoexpire I did check out your fork in the end. Very impressive! I see you have added a lot of functionality like windows WSL support which looks great. I was considering adding the check_commands functionality but in the end decided to remove it in order to keep this version more of a streamlined alternative. 

The `-4` option was a nice addition that I hadn't thought of.

timetoexpire commented on cloudflare-template.sh:
```diff
@@ -17,26 +17,20 @@ slackuri=""                                        # URI for Slack WebHook "http
-    ip=$(curl -s https://cloudflare.com/cdn-cgi/trace | grep -E '^ip' | sed -E 's/^ip=([0-9\.]*)$/\1/' || curl -s https://api.ipify.org || curl -s https://ipv4.icanhazip.com/)
+ipv4_regex='([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])'
+ip=$(curl -s -4 https://cloudflare.com/cdn-cgi/trace | grep -E '^ip'); ret=$?
+if [[ ! $ret == 0 ]]; then # In the case that cloudflare failed to return an ip.
```
Another way of doing it would be
`if [ ${#ip} -lt 10 ]; then`
It is not posble to have IPv4 "_ip=1.2.3.4_" with a string length less then 10.
Be also be able to remove `; ret=$?`

timetoexpire commented:
@9cco err, windows WSL support. It isn't working current one on github. An updated script coming out next mouth will have working. https://github.com/timetoexpire/cloudflare-ddns-updater/discussions/2

Microsoft has done something with WSL which I can not understand why they did what they done. But for whatever reason a fundamental part of Linux has been removed. So I have work around there actions. I dont think that I have been able to come up with my final solution to that problem. Should be able to use it in WSL in next update.

It other words, Microsoft why did do that, please explain? I am not totaly happy with my solution but it will work in WSL.

9cco commented:
I just discovered WSL myself. Considering implementing a backup solution using `rsync` with WSL from my win pc, but haven't poked at it yet. Sorry to hear they changed it for the worse! My partner works for MS, so maybe I can ask her 😄 I don't think she knows though. She's working with Azure so.

timetoexpire commented:
Alright, try this using (Debian/Ubuntu/Raspbian) that an virtual machine that isn't running using WSL or WSL2. Then use (Debian/Ubuntu) that using WSL or WSL2. Compare the output.
```
#!/bin/bash
echo "This is line 2"
logger "This is line 3"
echo "This is line 4"
logger -s "This is line 5"
echo "This is line 6"
tail /var/log/syslog
echo "This is line 8"
ls /dev/log -al
echo "This is line 9"
ls /var/log/syslog -al
echo "This is line 11"
```

PS Some versions of Linux that are not based upon Debian have logs in other locations. "This is line x" really to make it more obvious what going on when.

timetoexpire commented on cloudflare-template.sh:
```diff
@@ -17,17 +17,26 @@ slackuri=""                                        # URI for Slack WebHook "http
-  logger -s "DDNS Updater: No public IP found"
-  exit 1
+# Use regex to check for proper IPv4 format.
+if [[ ! $ip =~ $ipv4_regex ]]; then
```
This dosn't work, it an strange one but here simple test and soloution 
```ipv4_regex () {
  ip="$1"
  echo -n "IP [$ip] "
  if [[ ! $ip =~ ^$ipv4_regex$ ]]; then
    echo -n "[1] "
  fi
  if [[ ! $ip =~ $ipv4_regex ]]; then
    echo -n "[2] "
  fi
  echo " "
}

ipv4_regex='([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])\.([01]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])'

echo "My test triggered [1] your test has triggered [2]"
ipv4_regex 0.0.0.0
ipv4_regex 1.2.3.4
ipv4_regex 012.004.000.199
ipv4_regex 255.255.255.255
ipv4_regex 0.0.0.256
ipv4_regex 100.100.100.256
ipv4_regex 100.100.100.356
ipv4_regex 0.0.256.0
ipv4_regex 255.255.255.256
ipv4_regex 256.256.256.256
ipv4_regex 0.0.300.0
ipv4_regex 0..2.3
ipv4_regex 0.1
ipv4_regex 0.1f.2
ipv4_regex computer```

9cco replied:
Thank you for catching that! I think I just forgot to add the `^` and `$` in this regex as it should be there. Strange that it doesn't work, but yes it should be there. I implemented this change in the last commit.

K0p1-Git commented:
LGTM
